compiler/src/dmd/expressionsem.d

```diff
@@ -12543,6 +12547,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 ~ "implicitly coerces them to slices, "
                 ~ "which are compared by reference");
 
+        // https://issues.dlang.org/show_bug.cgi?id=23972
+        //
+        // `iface is obj`  ->  `cast(Object)iface is obj`
+        // `iface1 is iface2` ->  `cast(Object)iface1 is cast(Object)iface2`
+        if (!e1isNullExp && !e2isNullExp)
+        {
+            auto tc1 = exp.e1.type.isTypeClass();
+            auto tc2 = exp.e2.type.isTypeClass();
+            if (tc1 && tc2)
+            {
+                const id1 = tc1 && !tc1.sym.isCPPinterface() ? tc1.sym.isInterfaceDeclaration() : null;
+                const id2 = tc2 && !tc2.sym.isCPPinterface() ? tc2.sym.isInterfaceDeclaration() : null;
+                if (id1 || id2)
+                {
+                    if (id1 && id2)
```

Comment:
This looks weird. I would do it:

```d
if (id1)
{
    if (id2) { /* id1 && id2 */ }
    else { /* id1 */]
}
else if (id2)
{
    // id2
}
```
This gets rid of the extra condition of `if(id1 || id2)